diff --git a/gcc/config/aarch64/msys.h b/gcc/config/aarch64/msys.h
new file mode 100644
index 00000000000..04cd4618301
--- /dev/null
+++ b/gcc/config/aarch64/msys.h
@@ -0,0 +1,14 @@
+#ifndef GCC_AARCH64_MSYS_H
+#define GCC_AARCH64_MSYS_H
+
+/* Override cygwin default */
+#undef EXTRA_OS_CPP_BUILTINS
+#define EXTRA_OS_CPP_BUILTINS()        \
+  do {                                \
+    builtin_define("__CYGWIN__");      \
+    builtin_define("__MSYS__");        \
+    builtin_define("__MSYS2__");       \
+    builtin_define_std("unix");        \
+  } while (0)
+
+#endif /* GCC_AARCH64_MSYS_H */
